### PR TITLE
LIO contest loader

### PR DESCRIPTION
**Pull request by PPakalns**
Implemented simple LIO format loader based mainly on Italy (yaml) format and taking some ideas from polygon loader.  Implemented only features that are needed for lio. #12 

This can be run on contest configuration file or on task configuration file.
./cmscontrib/ImportContest.py -L Lio-contest contest_dir/contest.yaml
./cmscontrib/ImportTask.py -L Lio-task contest_dir/task_dir/task.yaml

Example configuration files for a part of this year warmup contest:
```yaml
name: warmup_test
description: 'Iesildīšanās kārta'
start: 2018-12-06T12:05:16+02:00
stop: 2018-12-06T19:05:16+02:00
per_user_time: 18000 # 5 * 3600
tasks:
  Palindr2:
    config: palindr/task.yaml
  Taisnst:
    config: taisnst/task.yaml
```
```yaml
name: 'Palindr2'
title: 'Palindroms'
statements:
  - ['../Uzdevumi/Skola2019_palindr2.pdf', 'lv']
public_groups: [0, 1, 2, 8, 15]
time_limit: 0.5
memory_limit: 256
```
```yaml
name: 'Taisnst'
title: 'Taisnstūri'
statements:
  - ['../Uzdevumi/Skola2019_taisnst2.pdf', 'lv']
time_limit: 0.7
memory_limit: 128
public_groups: [0, 1, 8, 16]
checker: ./checker/checker_jv.cpp
test_archive: ./testi/testi.zip
```

Some observations:
All contest, task imports of cms are intended to be run on the path that points to a directory. This creates a small issue for us because we often have the same task in two different contests. Additionally, we keep the tasks for each contest in the same directory. So we need to split them apart. Therefore I implemented loaders to work on the path to a config file instead of to the path to a directory.

The only issue with this is ContestLoader and `get_task_loader` already assumes that the provided task name is the real name of the task, therefore this creates a duplication in both config files: task configuration name field and contest configuration task name must be set to the same name. 

If we implemented this with separate directories then this duplication is simply kept in a different form (in the name of a task directory and still in the contest config file) And this still would rise issues with one task in two contests.

Remaining issue:

- [x] automatic newline converter to Unix newlines when extracting tests from zip archive. (Didn't find a nice way how to do this in python)  _This probably could be done better without loading all file in memory._

### CODE REVIEW

**Comment by PPakalns** [comment]
Successfully imported warmup2 contest.

My observations:
* The location and name of the test archive should be precisely defined. It differs between tasks in our repository.
* Options like:
  Minimum interval between submissions, minimum interval between user tests for contest. 
  Maximum number of submissions, maximum number of user tests for task.
  Should have default value in task loader.  (30, 30, 40, 40)
  So that we do not need to repeat them in contest.yaml, task.yaml configuration files.
* Timezone is not set to "Europe/Riga".
* Contest allowed localizations is not set to "lv".

All other settings are set up correctly.

**Comment by PPakalns** [comment]
Successfully imported regional contests that share one task. See svn for configuration file setup.

Additionally, tested the case when there are submissions and task configuration (tests, public groups) is changed.

For example, I saw this message:
```txt
 Publiski redzamais rezultāts:
10 / 6 
```

What happens. When a user sees the score or the detailed results for submission, he sees the results for the old dataset. Detailed results don't seem to leak any information about the new dataset, it shows detailed results for old dataset. (with scores for old public groups)

Therefore, when setting up contest **before the start**, we can execute the import multiple times. After each import we should reevaluate all solutions for affected tasks (use task importer if necessary). 

In conclusion, the configuration part of contests and tasks before the start is moved to these config files.  We can keep version history of them.

File: cmscontrib/loaders/lio.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+            trg_dict[key] = conv(src_dict[key])
+        else:
+            trg_dict[key] = src_dict[key]
+    elif default is not None:
+        trg_dict[key] = default
+
+
+class LioTaskLoader(TaskLoader):
+
+    short_name = "Lio-task"
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Other loaders follow the Python variable naming convention here. I suggest `lio_task` and `lio_contest`.

File: cmscontrib/loaders/lio.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    if key in src_dict:
+        if conv:
+            trg_dict[key] = conv(src_dict[key])
+        else:
+            trg_dict[key] = src_dict[key]
+    elif default is not None:
+        trg_dict[key] = default
+
+
+class LioTaskLoader(TaskLoader):
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py:52]: The Italy YAML loader derives simultaneously from `TaskLoader`, `ContestLoader` … etc. Could we do the same and just have `LioLoader`?

File: cmscontrib/loaders/lio.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+from cms import TOKEN_MODE_DISABLED, TOKEN_MODE_FINITE, TOKEN_MODE_INFINITE
+from cmscommon.constants import \
+    SCORE_MODE_MAX, SCORE_MODE_MAX_SUBTASK, SCORE_MODE_MAX_TOKENED_LAST
+from .italy_yaml import load_yaml_from_path, make_timedelta
+from datetime import timedelta
+
+logger = logging.getLogger(__name__)
+
+
+def set_if_present(src_dict, trg_dict, key, conv=None, default=None):
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Maybe the default value for `conv` can be `lambda x: x`.

File: cmscontrib/loaders/lio.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        self.conf = load_yaml_from_path(self.path)
+
+
+    @staticmethod
+    def detect(path):
+        # TODO: Support auto detection
+        return False
+
+
+    def parse_point_file(self, point_path):
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py:70]: Ideally we would specify subgroup points in the YAML file instead of parsing a text file, but as a first step this is good too.

I suggest to add a TODO here to replace this with YAML loading.

File: cmscontrib/loaders/lio.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        args['submission_format'] = [f"{name}.%l"]
+
+        set_if_present(self.conf, args, 'score_precision')
+
+        score_mode = self.conf.get('score_mode', SCORE_MODE_MAX_TOKENED_LAST)
+        if score_mode in [SCORE_MODE_MAX, SCORE_MODE_MAX_SUBTASK, SCORE_MODE_MAX_TOKENED_LAST]:
+            args['score_mode'] = score_mode
+        else:
+            logger.critical("Unknown score mode provided")
+            raise ValueError
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Some places in this method raise an exception and some others return `None` on error. Which is the expected way?

File: cmscontrib/loaders/lio.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+        if 'checker' in self.conf:
+            logger.info("Checker found, compiling")
+            checker_src = os.path.join(self.task_dir, self.conf['checker'])
+            checker_exe = os.path.join(os.path.dirname(checker_src), "checker")
+            if config.installed:
+                testlib_path = "/usr/local/include/cms"
+            else:
+                testlib_path = os.path.join(os.path.dirname(__file__), "polygon")
+            code = subprocess.call(["g++", "-x", "c++", "-O2", "-static",
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: I suggest to add also `-pipe` and `-s`.

File: cmscontrib/loaders/lio.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        input_filename = self.conf.get("input_filename", "")
+        output_filename = self.conf.get("output_filename", "")
+
+        # No grader support
+        compilation_param = "alone"
+
+        if 'checker' in self.conf:
+            logger.info("Checker found, compiling")
+            checker_src = os.path.join(self.task_dir, self.conf['checker'])
+            checker_exe = os.path.join(os.path.dirname(checker_src), "checker")
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Here we modify the source file hierarchy, which feels a bit wrong.

If you have interest, you could try replacing this with `tempfile.NamedTemporaryFile` (the file name for GCC call can be obtained with `f.name`) and using `put_file_from_fobj` later.

Otherwise, you can leave this as is :)

File: cmscontrib/loaders/lio.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+                checker_exe, "Checker for task {name}"
+            )
+            args["managers"]["checker"] = Manager("checker", digest)
+            evaluation_param = "comparator"
+        else:
+            evaluation_param = "diff"
+
+        point_file = os.path.join(self.task_dir, self.conf.get('point_file', 'punkti.txt'))
+        points_per_group = self.parse_point_file(point_file)
+        if sum(points_per_group.values()) != 100:
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Should this be moved into `parse_point_file`?

File: cmscontrib/loaders/lio.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+            # Extract them in correct order and update subtask list and testcases
+            for group in sorted(test_files.keys()):
+                for test_in_group in sorted(test_files[group].keys()):
+                    tests_per_group[group] += 1
+                    testcase = test_files[group][test_in_group]
+                    if 'input' not in testcase or 'output' not in testcase:
+                        logger.critical(f"Input or output file not found for test {group}{test_in_group}")
+                        return None
+                    with zip.open(testcase['input'], 'r') as input_file:
+                        content = io.TextIOWrapper(input_file, encoding='ascii', newline=None).read()
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py:226]: Here and below we are loading a potentially huge file in memory. Can we only use file objects instead?

**Comment by andreyv** [comment]
Running `cmsImportContest -L Lio-contest 2_novads/contest_junior.yaml`, I get
```
2019-07-22 20:33:57,936 - INFO [<unknown>] Using configuration file /etc/cms.conf.
2019-07-22 20:33:58,379 - INFO [<unknown>] Loading parameters for contest junior.
2019-07-22 20:33:58,380 - INFO [<unknown>] Contest parameters loaded.
2019-07-22 20:33:58,394 - INFO [<unknown>] Creating contest on the database.
2019-07-22 20:33:58,407 - ERROR [<unknown>] Task "Mazakais" not found in database. Use --import-task to import it.
2019-07-22 20:33:58,407 - INFO [<unknown>] Error while importing, no changes were made.
```

Is it supposed to work this way?

File: cmscontrib/loaders/lio.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+                        input_digest = self.file_cacher.put_file_content(
+                            content.encode('ascii'),
+                            f"Input {testcase['input']} for task {task.name}")
+                    with zip.open(testcase['output'], 'r') as output_file:
+                        content = io.TextIOWrapper(output_file, encoding='ascii', newline=None).read()
+                        output_digest = self.file_cacher.put_file_content(
+                            content.encode('ascii'),
+                            f"Output {testcase['output']} for task {task.name}")
+                    codename = f"{group:03}{test_in_group}"
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: What do you think about lowering this to two digits? I understand the intent, but it should be rare, and seeing 001 002 003 … in the testcase list makes it harder to read, at least for me.

File: cmscontrib/loaders/lio.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+                logger.info(f"Loading statement: {statement}")
+                digest = self.file_cacher.put_file_from_path(
+                    os.path.join(self.task_dir, path),
+                    f"Statement for task {name} (lang: {lang})",
+                )
+                args['statements'][lang] = Statement(lang, digest)
+            if args['statements']:
+                args['primary_statements'] = self.conf.get('primary_statements', ["lv"])
+
+        args['submission_format'] = [f"{name}.%l"]
```
Comment (andreyv) [inline cmscontrib/loaders/lio.py]: Note that this doesn't necessarily match the task name. Previously we had, for example, tasks named `KastesJ` and `KastesV`, but the expected submission format and Java class name was `Kastes`. I suppose this necessitates one more field in the YAML file.

**Review by andreyv — CHANGES_REQUESTED** [review]
I really like how the newlines are converted :)

**Comment by PPakalns** [comment]
> Running `cmsImportContest -L Lio-contest 2_novads/contest_junior.yaml`, I get
> 
> ```
> 2019-07-22 20:33:57,936 - INFO [<unknown>] Using configuration file /etc/cms.conf.
> 2019-07-22 20:33:58,379 - INFO [<unknown>] Loading parameters for contest junior.
> 2019-07-22 20:33:58,380 - INFO [<unknown>] Contest parameters loaded.
> 2019-07-22 20:33:58,394 - INFO [<unknown>] Creating contest on the database.
> 2019-07-22 20:33:58,407 - ERROR [<unknown>] Task "Mazakais" not found in database. Use --import-task to import it.
> 2019-07-22 20:33:58,407 - INFO [<unknown>] Error while importing, no changes were made.
> ```
> 
> Is it supposed to work this way?

When importing contest you need to pass the relevant flags to define the behaviour for importer:
```
  -i, --import-tasks    import tasks if they do not exist
  -u, --update-contest  update an existing contest
  -U, --update-tasks    update existing tasks
  -S, --no-statements   do not import / update task statements
  --delete-stale-participations
                        when updating a contest, delete the participations not
                        in the new contest, including their submissions and
                        other data
```

File: cmscontrib/loaders/lio.py (hunk 2 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py:52]: I would like to keep them separated as their functionality doesn't overlap. It improves readability and structure of code.

File: cmscontrib/loaders/lio.py (hunk 4 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py:70]: I agree, but we probably need to introduce this format for importing tasks first and then we can ask other members to switch their group point definition way to the yaml file and it would be nice if they would write the yaml file too. (We probably still will need to fix it) At least now it is more simple for them.
But if we standardize the layout of files there wouldn't be a lot of fields to specify for them.

File: cmscontrib/loaders/lio.py (hunk 5 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: I confused this with `get_task`, `get_contest` and similar functions that accepts returned None, will fix this.

Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: italy loader were ignoring errors in this way. I used the same behaviour in the most part. Will replace with correctly raised exceptions where necessary.

File: cmscontrib/loaders/lio.py (hunk 7 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: With NamedTemporaryFile there is platform dependant issues about if file can be opened the second time, I will use temporary directory instead.

File: cmscontrib/loaders/lio.py (hunk 9 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py:226]: I wanted to do the new line formatting through the streams and objects before, but I remember that I didn't find how to do that. I remember that  TextIOWrapper didn't support that. Will look into it when I will have time, but as RAM memory is huge nowadays I don't think that these small tests will raise any issue.

File: cmscontrib/loaders/lio.py (hunk 8 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: Done

File: cmscontrib/loaders/lio.py (hunk 11 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: Added field.

File: cmscontrib/loaders/lio.py (hunk 10 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: Modified it to automatically use the minimum necessary digit count.

File: cmscontrib/loaders/lio.py (hunk 3 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: Applied.

File: cmscontrib/loaders/lio.py (hunk 1 above)
Comment (PPakalns) [reply cmscontrib/loaders/lio.py]: Applied.

**Review by andreyv — APPROVED** [review]

**Comment by andreyv** [comment]
I think this is good now, thanks :)

The diff shows some unrelated changes, can you please fix this before merge?
